src/viam/sdk/tests/test_navigation.cpp

```diff
@@ -0,0 +1,136 @@
+#define BOOST_TEST_MODULE test module test_navigation
+
+#include <boost/test/included/unit_test.hpp>
+
+#include <viam/sdk/services/navigation.hpp>
+#include <viam/sdk/tests/mocks/mock_navigation.hpp>
+#include <viam/sdk/tests/test_utils.hpp>
+
+BOOST_TEST_DONT_PRINT_LOG_VALUE(viam::sdk::Navigation::Mode)
+BOOST_TEST_DONT_PRINT_LOG_VALUE(viam::sdk::Navigation::Path)
+BOOST_TEST_DONT_PRINT_LOG_VALUE(viam::sdk::Navigation::LocationResponse)
+BOOST_TEST_DONT_PRINT_LOG_VALUE(std::vector<std::string>)
+BOOST_TEST_DONT_PRINT_LOG_VALUE(std::vector<viam::sdk::Navigation::Path>)
+BOOST_TEST_DONT_PRINT_LOG_VALUE(std::vector<viam::sdk::geo_point>)
+BOOST_TEST_DONT_PRINT_LOG_VALUE(viam::sdk::Navigation::MapType)
+
+namespace viam {
+namespace sdktests {
+namespace navigation {
+
+using namespace viam::sdk;
+
+BOOST_AUTO_TEST_CASE(nav_mode) {
+    auto mock = std::make_shared<MockNav>("mock_nav");
+    client_to_mock_pipeline<Navigation>(mock, [&](Navigation& client) {
+        client.set_mode(Navigation::Mode::k_manual);
+        BOOST_CHECK_EQUAL(client.get_mode(), Navigation::Mode::k_manual);
+        client.set_mode(Navigation::Mode::k_explore);
+        BOOST_CHECK_EQUAL(client.get_mode(), Navigation::Mode::k_explore);
+    });
+}
+
+BOOST_AUTO_TEST_CASE(nav_get_location) {
+    auto mock = std::make_shared<MockNav>("mock_nav");
+    client_to_mock_pipeline<Navigation>(mock, [&](Navigation& client) {
+        auto loc = client.get_location();
+        BOOST_CHECK_EQUAL(loc, Navigation::LocationResponse{});
+        mock->loc.compass_heading = 1;
+        loc = client.get_location();
+        BOOST_CHECK_CLOSE(loc.compass_heading, 1., 1e-4);
+    });
+}
+
+// weak backport of std::transform.
+template <typename T, typename U>
```

Comment:
I think just template this as `template <typename T, typename U, typename Func>` and have the second argument be `Func f`

Reply:
that made it sad

could specify template param I guess? but can I leave as-is?

> /host/src/viam/sdk/tests/test_navigation.cpp: In lambda function:
> /host/src/viam/sdk/tests/test_navigation.cpp:78:14: error: no matching function for call to 'mapOver(std::vector<viam::sdk::Navigation::Waypoint>&, std::__cxx11::basic_string<char> (*)(const viam::sdk::Navigation::Waypoint&))'
>    78 |             });
>       |              ^
> /host/src/viam/sdk/tests/test_navigation.cpp:46:16: note: candidate: 'template<class T, class U, class Func> std::vector<ValueType> viam::sdktests::navigation::mapOver(const std::vector<U>&, Func)'
>    46 | std::vector<T> mapOver(const std::vector<U>& source, Func f) {
>       |                ^~~~~~~
> /host/src/viam/sdk/tests/test_navigation.cpp:46:16: note:   template argument deduction/substitution failed:
> /host/src/viam/sdk/tests/test_navigation.cpp:78:14: note:   couldn't deduce template parameter 'T'
>    78 |             });